rafthttp/snapshot.go

```diff
@@ -0,0 +1,188 @@
+// Copyright 2015 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rafthttp
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/coreos/etcd/Godeps/_workspace/src/golang.org/x/net/context"
+	"github.com/coreos/etcd/pkg/types"
+	"github.com/coreos/etcd/raft"
+	"github.com/coreos/etcd/raft/raftpb"
+)
+
+// timeout that waits snapshot to be called to write out
+// Before snapshot is called to write out, raft snapshot message needs to be sent out,
+// then let remote member send request to ask for snapshot, so at least
+// two maximal RTT is required, which is 10s.
+const snapshotWaitTimeout = 10 * time.Second
+
+type urlPickerGetter interface {
+	getURLPicker(id types.ID) *urlPicker
+}
+
+func newSnapshotBuffer(r Raft, snapst SnapshotStore) *snapshotBuffer {
+	return &snapshotBuffer{
+		r:      r,
+		snapst: snapst,
+	}
+}
+
+func newSnapshotProcessor(tr http.RoundTripper, upGetter urlPickerGetter, r Raft, snapst SnapshotStore) *snapshotProcessor {
+	return &snapshotProcessor{
+		tr:       tr,
+		upGetter: upGetter,
+		r:        r,
+		snapst:   snapst,
+	}
+}
+
+// snapshotBuffer represents a buffer of snapshot with write method.
+type snapshotBuffer struct {
+	r      Raft
+	snapst SnapshotStore
+
+	// protect fields below
+	mu sync.Mutex
+	// whether it has snapshot
+	hasSnap bool
+	// the member ID that the snapshot is sent to
+	snapTo types.ID
+	// the index of the snapshot
+	snapIndex uint64
+	// when the timer expires, it will call the function attached via AfterFunc
+	// to close the snapshot
+	snapCloseTimer *time.Timer
+}
+
+// addSnap adds snapshot into snapshot buffer.
```

Comment:
It adds the snapshot at https://github.com/coreos/etcd/pull/3648/files#diff-15a30f78ad76cc19daa724a361c77fc2R86 .